WordPress/openverse
# Run the app as ASGI

## Fixes



Fixes #2790 by @sarayourfriend 

## Description


In #3000, I've tried to implement the async client sharing necessary for #2788. The idea of that issue is to do async client instantiation "the right way". However, in that PR, I've discovered the deep complexities of trying to do things this way. There is, in fact, no way to do things the right way in an "async under WSGI" approach that we wouldn't immediately remove after we switched the app to run under ASGI.

On the other hand, switching the app to run under ASGI, without making other changes, does not create any such issues. Django automatically wraps sync views in `sync_to_async` and our application runs perfectly (the integration tests show that).

It does, however, require making some relatively significant changes to the way we deploy the application. Namely, that is to switch to uvicorn and remove gunicorn in favour of scaling our ECS tasks with smaller individual task resources. In other words, rather than having tasks with more resources split their resources with four gunicorn workers, we'll have tasks with fewer individual resources, and scale the number of "workers" by increasing the number of tasks ECS provisions. This removes the gunicorn worker-manager middleman and relies on the orchestrator (ECS in our case) to create the correct number of "workers". Indeed, FastAPI's deployment advice recommends this approach for containerised applications outside of special cases (which I don't think we are): https://fastapi.tiangolo.com/deployment/docker/#one-process-per-container

So, with all of that in mind, this PR does the following:

1. Run the application using uvicorn and convert gunicorn configuration to uvicorn
2. Implement an ASGI handler class that adds lifespan support to Django (compensates for https://code.djangoproject.com/ticket/31508)

This PR will require some changes to the API infrastructure deployment. Specifically, we'll want to do the following before deploying this PR:

1. Increase the number of tasks for the API
2. Then deploy this PR
3. Then scale the task resources down and redeploy

For scaling the resources down, we need to consider the existing settings:

5 tasks with 1 vCPU and 4 GB RAM each.

Each task runs 4 workers. To get the exact same number of workers, we'd need 20 tasks. If we scale each part by a quarter, that is 0.25 vCPU and 1 GB RAM each. Surprising, even with the high number of individual tasks, that puts us at 1k USD less per 30-day period than we spend now.

That being said, I don't think this mathematical approach is going to be reliable. Instead, we should consider that we do not fully exercise our API instances to the maximum capacity. That means there could be room to deploy fewer tasks with more resources or just fewer tasks overall. So long as we match roughly the same worker-to-resource allocation ratio we have now, the price differences are either negligible or offer significant savings.

## Testing Instructions

Checkout this branch and run `just build web` to get the new dependencies. Then run `just api/up` and make requests to the app. Check the logs with `just logs web` and confirm they look good and match the expected output patterns for our logging configuration (it should just match `main`).

## Checklist


- [N/A] I added or updated tests for the changes I made (if applicable).
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
